### PR TITLE
Added new CLI named clusterkubevirtadm, to manage the infra cluster

### LINKED ISSUE #106: Add clusterkubevirtadm utility to create ServiceAccount for infrastructure resources creation
Create new Cobra CLI utility
Name: clusterkubevirtadm
It create will ServiceAccount in the Infrastructure clusture to be used for all the resources creation
It will also create a Role to this ServiceAccount, include minimal permissions required for the resources creation
It will output the credentials file (kubeconfig) for using the new service account

This utility, will allow the infrastructure admin create special credentials for each tenant cluster creation.

### PR DESCRIPTION
**Pull request by nunnatsa**
Added a new CLI tool to manage the infra cluster. For now, the tool supports two commands:

1. `clusterkubevirtadm create credentials`
    this command creates a namespace, serviceAccount, role and roleBinding in the infra cluster. If the namespace is already exist. the command returns error.
2. `clusterkubevirtadm apply credentials`
    this command creates a namespace, serviceAccount, role and roleBinding in the infra cluster. If the namespace is already exist. the command will update the resources to the latest version.
3. `clusterkubevirtadm get credentials`
    this command generates a kubeconfig file to use the serviceAccount created with the first command.

for all the three commans above, the `cred` and `creds` aliases can replace `credentials`.

The build the new tool, use:
* `make clusterkubevirtadm-linux` - for linux application
* `make clusterkubevirtadm-macos` - for macOS application
* `make clusterkubevirtadm-win` - for windows application
* `make clusterkubevirtadm-all` - for all the above + run the unit tests
* `make clusterkubevirtadm-test` - to run the unit tests

Fixes #106

Signed-off-by: Nahshon Unna-Tsameret <nunnatsa@redhat.com>

**Release notes**:

```release-note
Added new CLI named clusterkubevirtadm, to manage the infra cluster
```

### CODE REVIEW

**Review by nirarg — COMMENTED** [review]
Thank you @nunnatsa this is great progress for having the utility
I have few thoughts that I think should be discussed:
1. I'm not sure that the parameter `cluster-name` is the right name, as there is no guaranty that the user will actually use this `cluster-name` for the actual cluster creation
2. Should the credentials create cmd fail if the namespace is already exists? maybe the behavior should be different in case the user provide the namespace?

**Comment by nirarg** [comment]
I think that https://github.com/kubernetes-sigs/cluster-api-provider-kubevirt/blob/9b27dbc6f1f07d4b2f3271c144cd3db585d9e1ea/clusterkubevirtadm/cmd/credentials/create.go should be moved to be under `clusterkubevirtadm/cmd/create/credentials` - in order to make the hyrarchy more clear
Similar comment for `get` operation

**Comment by nirarg** [comment]
Need to change e2e-tests to use this new cmd
https://github.com/kubernetes-sigs/cluster-api-provider-kubevirt/blob/main/e2e-tests/create-cluster_test.go

File: clusterkubevirtadm/cmd/credentials/credentials.go [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+	return &rbacv1.Role{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      roleName,
+			Namespace: opts.Namespace,
+		},
+		Rules: []rbacv1.PolicyRule{
+			{
+				APIGroups: []string{rbacv1.APIGroupAll},
+				Resources: []string{rbacv1.ResourceAll},
+				Verbs:     []string{rbacv1.VerbAll},
```
Comment (davidvossel) [inline clusterkubevirtadm/cmd/credentials/credentials.go:157]: i think we can limit this to `get,create,delete` for secrets and `*` for VirtualMachines

File: clusterkubevirtadm/cmd/credentials/credentials.go [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	clientcmdlatest "k8s.io/client-go/tools/clientcmd/api/latest"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	CmdParamClusterName        = "cluster-name"
+	CmdParamServiceAccountName = "service-account-name"
+	CmdParamServiceNamespace   = "namespace"
```
Comment (davidvossel) [inline clusterkubevirtadm/cmd/credentials/credentials.go:42]: lets' just start with the namespace parameter

File: clusterkubevirtadm/cmd/create/credentials/create_credentials.go [hunk 3]
```diff
@@ -41,37 +39,11 @@ Exists with error if the namespace is already exist.
 `,
 		SilenceUsage: true,
 		RunE: func(cmd *cobra.Command, args []string) error {
-			return create(cmd.Context(), opts)
+			return common.Apply(cmd.Context(), opts)
```
Comment (nirarg) [inline clusterkubevirtadm/cmd/create/credentials/create_credentials.go:42]: Need to validate the namespace doesn't exist before calling to `common.Apply`

File: clusterkubevirtadm/cmd/credentials/credentials.go [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+		} else {
+			return fmt.Errorf("can't get the serviceAccount; %w", err)
+		}
+	}
+
+	fmt.Println("Found ServiceAccount", common.ServiceAccountName)
+	return nil
+}
+
+func createServiceAccount(ctx context.Context, cmdCtx cmdContext) error {
```
[the remark this answers is not in the document]
Comment (nunnatsa) [reply clusterkubevirtadm/cmd/credentials/credentials.go:97]: not the same. this function actually create the SA in the cluster. the other ones only create an object.

File: clusterkubevirtadm/cmd/credentials/credentials.go [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+		},
+		RoleRef: rbacv1.RoleRef{
+			APIGroup: rbacv1.GroupName,
+			Kind:     "Role",
+			Name:     roleName,
+		},
+	}
+}
+
+func createOrUpdatePermissions(ctx context.Context, cmdCtx cmdContext, co clientOperation) error {
```
Comment (nirarg) [inline clusterkubevirtadm/cmd/credentials/credentials.go:201]: Suggest name: `createOrUpdateResources`

**Review by nirarg — COMMENTED** [review]
Two more minor comments
I think this is ready to be pushed
Please squash the commits before

**Comment by nirarg** [comment]
Also need to add quick start section to the repo README, describe how to build and use this utility
This can be done in different PR

**Pull request merged**